File: apps/driver-android/src/main/kotlin/lk/mageride/driver/wallet/WalletScreen.kt
package lk.mageride.driver.wallet

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lk.mageride.driver.R
import lk.mageride.driver.ui.MoneyFormat
import lk.mageride.driver.ui.component.DashboardBanner
import lk.mageride.driver.ui.component.LabelledTextField
import lk.mageride.driver.ui.component.MageRideCta
import lk.mageride.driver.ui.component.SectionLabel
import lk.mageride.driver.ui.labelRes
import lk.mageride.driver.ui.theme.MageRideTheme
import lk.mageride.shared.domain.wallet.WalletAlert
import org.koin.androidx.compose.koinViewModel

/**
 * **SCR-DA-021 · wallet & fee status** (US-9.7, US-9.1, US-9.9).
 *
 * The wireframe, top to bottom: a `Wallet` app bar with no back arrow — this is bottom-nav tab 3,
 * not a pushed screen — the read-only balance in display type, the daily-fee card with its three
 * `kv` rows, and the two button rows that open the other four screens in this group.
 *
 * **One thing here is not in the wireframe: the *"Warn me below Rs 200"* row.** D2' §SCR-DA-021's
 * own states line calls the threshold *"driver-set"* and the C073 deliverable asks for the setting,
 * and there is nowhere else in this group it could live. [WalletPreferences] explains why it is
 * stored on the handset, and the dialog says so to the driver rather than implying a server-side
 * change. Recorded in the C073 handoff.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun WalletScreen(
    onTopUp: () -> Unit,
    onRequestCredit: () -> Unit,
    onTransferCredit: () -> Unit,
    onHistory: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val viewModel: WalletViewModel = koinViewModel()
    val state by viewModel.state.collectAsStateWithLifecycle()
    var editingThreshold by remember { mutableStateOf(false) }

    Scaffold(
        modifier = modifier.fillMaxSize(),
        topBar = { TopAppBar(title = { Text(text = stringResource(R.string.wallet_title)) }) },
    ) { insets ->
        Column(
            modifier = Modifier
                .padding(insets)
                .fillMaxSize()
                .verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
        ) {
            if (state.loading) {
                LinearProgressIndicator(modifier = Modifier.fillMaxWidth())
            }
            state.error?.let { message ->
                DashboardBanner(text = stringResource(message), accent = MaterialTheme.colorScheme.error)
            }
            WalletNotice(state = state)

            Column(
                modifier = Modifier.padding(horizontal = MageRideTheme.spacing.md),
                verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
            ) {
                BalanceBlock(state = state)
                DailyFeeCard(state = state)
                ThresholdRow(state = state, onEdit = { editingThreshold = true })
                WalletActions(
                    onTopUp = onTopUp,
                    onRequestCredit = onRequestCredit,
                    onTransferCredit = onTransferCredit,
                    onHistory = onHistory,
                )
            }
        }
    }

    if (editingThreshold) {
        ThresholdDialog(
            current = state.threshold.amountMinor,
            onSave = { minor ->
                viewModel.setThreshold(minor)
                editingThreshold = false
            },
            onReset = {
                viewModel.clearThreshold()
                editingThreshold = false
            },
            onDismiss = { editingThreshold = false },
        )
    }
}

/**
 * The wireframe's centred *"Balance (read-only)"* and its display figure.
 *
 * `balance`, not `available`: US-9.7 calls this the balance and D2' marks it read-only. The
 * spendable figure is what every *decision* in this group is checked against — see
 * [WalletState.belowDayFee] and `CreditTransferRules.rejectionFor` — and it is shown as its own
 * line only when the two differ, because a driver with no accrued debt should not be reading two
 * numbers for one wallet.
 */
@Composable
private fun BalanceBlock(state: WalletState, modifier: Modifier = Modifier) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .padding(top = MageRideTheme.spacing.sm),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Text(
            text = stringResource(R.string.wallet_balance_label),
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        Text(
            text = state.balance?.let { MoneyFormat.rupees(it) } ?: MoneyFormat.EMPTY,
            style = MaterialTheme.typography.displaySmall,
            color = MaterialTheme.colorScheme.onSurface,
        )

        val debt = state.standing.standing?.outstandingDebt
        if (debt != null && debt.amountMinor > 0) {
            Text(
                text = stringResource(R.string.wallet_available_after_debt, MoneyFormat.rupees(debt)),
                style = MaterialTheme.typography.labelSmall,
                color = MageRideTheme.status.warning,
                textAlign = TextAlign.Center,
            )
        }
    }
}

/** The wireframe's `Daily fee` card — vehicle, rate, and today's status (US-9.1/9.7). */
@Composable
private fun DailyFeeCard(state: WalletState, modifier: Modifier = Modifier) {
    Card(
        modifier = modifier.fillMaxWidth(),
        shape = RoundedCornerShape(MageRideTheme.radius.md),
        colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surfaceVariant),
        elevation = CardDefaults.cardElevation(defaultElevation = MageRideTheme.elevation.level0),
    ) {
        Column(
            modifier = Modifier.padding(MageRideTheme.spacing.sm),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xxs),
        ) {
            SectionLabel(text = stringResource(R.string.wallet_daily_fee))

            val fee = state.standing.dailyFee
            if (fee == null) {
                Text(
                    text = stringResource(R.string.wallet_fee_unavailable),
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                return@Column
            }

            FeeRow(
                label = stringResource(R.string.wallet_fee_vehicle),
                value = stringResource(fee.vehicleType.labelRes()),
            )
            FeeRow(
                label = stringResource(R.string.wallet_fee_rate),
                value = stringResource(
                    R.string.wallet_fee_rate_value,
                    state.standing.dailyRate?.let { MoneyFormat.rupees(it) } ?: MoneyFormat.EMPTY,
                ),
            )

            // "PAID ✓ (1st trip free)" is two facts in one line, and the qualifier only belongs on
            // a day the free trip has not been spent — see `WalletFeeStanding.firstTripStillFree`.
            val status = stringResource(
                if (state.standing.feePaid) R.string.wallet_fee_paid else R.string.wallet_fee_unpaid,
            )
            FeeRow(
                label = stringResource(R.string.wallet_fee_today),
                value = if (state.standing.firstTripStillFree) {
                    stringResource(R.string.wallet_fee_status_first_free, status)
                } else {
                    status
                },
                accent = if (state.standing.feePaid) MageRideTheme.status.success else MageRideTheme.status.warning,
            )
        }
    }
}

/** The wireframe's `kv` — a muted label on the left, the value on the right. */
@Composable
private fun FeeRow(label: String, value: String, modifier: Modifier = Modifier, accent: Color? = null) {
    Row(
        modifier = modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
    ) {
        Text(
            text = label,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.weight(1f),
        )
        Text(
            text = value,
            style = MaterialTheme.typography.bodyMedium,
            color = accent ?: MaterialTheme.colorScheme.onSurface,
        )
    }
}

/**
 * The two banner states, ranked hardest first.
 *
 * Overdrawn is D5' §9.4's *"Top Up Required"*, *"below one day's fee"* is D2' §SCR-DA-021's, and
 * the low-balance nudge is the softest of the three — see [WalletState.belowDayFee] on why the
 * first two are different questions rather than one restated.
 */
@Composable
private fun WalletNotice(state: WalletState, modifier: Modifier = Modifier) {
    when {
        state.alert is WalletAlert.TopUpRequired -> DashboardBanner(
            modifier = modifier,
            text = stringResource(
                R.string.wallet_top_up_required,
                MoneyFormat.rupees((state.alert as WalletAlert.TopUpRequired).owed),
            ),
            accent = MaterialTheme.colorScheme.error,
        )

        state.belowDayFee -> DashboardBanner(
            modifier = modifier,
            text = stringResource(
                R.string.wallet_below_day_fee,
                state.standing.dailyRate?.let { MoneyFormat.rupees(it) } ?: MoneyFormat.EMPTY,
            ),
            accent = MaterialTheme.colorScheme.error,
        )

        state.alert is WalletAlert.LowBalance -> DashboardBanner(
            modifier = modifier,
            text = stringResource(R.string.wallet_low_balance, MoneyFormat.rupees(state.threshold)),
            accent = MageRideTheme.status.warning,
        )

        else -> Unit
    }
}

/** The wireframe's two `btn-row`s — one CTA and three outlined actions. */
@Composable
private fun WalletActions(
    onTopUp: () -> Unit,
    onRequestCredit: () -> Unit,
    onTransferCredit: () -> Unit,
    onHistory: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
    ) {
        MageRideCta(label = stringResource(R.string.wallet_action_top_up), onClick = onTopUp)

        Row(horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs)) {
            OutlinedButton(onClick = onRequestCredit, modifier = Modifier.weight(1f)) {
                Text(text = stringResource(R.string.wallet_action_request))
            }
            OutlinedButton(onClick = onTransferCredit, modifier = Modifier.weight(1f)) {
                Text(text = stringResource(R.string.wallet_action_transfer))
            }
        }
        OutlinedButton(onClick = onHistory, modifier = Modifier.fillMaxWidth()) {
            Text(text = stringResource(R.string.wallet_action_history))
        }
    }
}

/** *"Warn me below Rs 200"* — the driver's own low-balance line (US-9.9). */
@Composable
private fun ThresholdRow(state: WalletState, onEdit: () -> Unit, modifier: Modifier = Modifier) {
    TextButton(onClick = onEdit, modifier = modifier) {
        Text(
            text = stringResource(R.string.wallet_threshold_row, MoneyFormat.rupees(state.threshold)),
            style = MaterialTheme.typography.labelLarge,
        )
    }
}

/**
 * The threshold dialog.
 *
 * The body says the setting lives on this handset, because it does and because a driver who
 * believed they had changed when MageRide texts them would be misled — the `LOW_BALANCE` push runs
 * on the admin's figure and there is no route that would let this one reach it.
 */
@Composable
private fun ThresholdDialog(current: Long, onSave: (Long) -> Unit, onReset: () -> Unit, onDismiss: () -> Unit) {
    var typed by remember { mutableStateOf(WalletInput.rupeesOf(current)) }
    val amountMinor = WalletInput.amountMinor(typed)

    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text(text = stringResource(R.string.wallet_threshold_title)) },
        text = {
            Column(verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs)) {
                Text(
                    text = stringResource(R.string.wallet_threshold_body),
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                LabelledTextField(
                    label = stringResource(R.string.wallet_amount_label),
                    value = typed,
                    onValueChange = { typed = WalletInput.rupeeDigits(it) },
                    prefix = MoneyFormat.PREFIX,
                )
            }
        },
        confirmButton = {
            TextButton(onClick = { amountMinor?.let(onSave) }, enabled = amountMinor != null) {
                Text(text = stringResource(R.string.action_save))
            }
        },
        dismissButton = {
            TextButton(onClick = onReset) { Text(text = stringResource(R.string.wallet_threshold_reset)) }
        },
    )
}

File: apps/driver-ios/DriverApp/Wallet/WalletScreen.swift
import MageRideShared
import SwiftUI

/// **SCR-DI-021 · wallet & fee status** (US-9.7, US-9.1, US-9.9).
///
/// The wireframe, top to bottom: the large title *"Wallet"* — this is tab 3, not a pushed screen, so
/// there is no back button — the read-only balance in display type, the daily-fee card with its three
/// `kv` rows, and the two `btn-row`s that open the other four screens in this group.
///
/// **One thing here is not in the wireframe: the *"Warn me below Rs 200"* row.** D2' §SCR-DI-021's own
/// states line calls the threshold *"driver-set"* and the C091 deliverable asks for the setting, and
/// there is nowhere else in this group it could live. ``WalletPreferences`` explains why it is stored
/// on the handset, and the sheet says so to the driver rather than implying a server-side change.
/// Recorded in the C091 handoff.
///
/// `@MainActor` on the whole view, not on its initialiser — see ``ProfileSetupScreen`` for why.
@MainActor
struct WalletScreen: View {

    @StateObject private var model: WalletModel

    private let onOpen: (DriverRoute) -> Void

    @State private var isEditingThreshold = false

    init(model: @autoclosure @escaping () -> WalletModel, onOpen: @escaping (DriverRoute) -> Void) {
        _model = StateObject(wrappedValue: model())
        self.onOpen = onOpen
    }

    var body: some View {
        ScrollView {
            VStack(spacing: MageRideSpacing.sm) {
                notice

                if let errorKey = model.state.errorKey {
                    DashboardBanner(text: errorKey.localised, accent: MageRideColor.error)
                        .onTapGesture(perform: model.dismissError)
                }

                balance
                dailyFeeCard
                thresholdRow
                actions
            }
            .padding(MageRideSpacing.md)
        }
        .frame(maxWidth: .infinity)
        .background(MageRideColor.background)
        .navigationTitle(Text(key: "wallet_title"))
        .navigationBarTitleDisplayMode(.large)
        .refreshable { await model.refresh() }
        .task { await model.refresh() }
        .sheet(isPresented: $isEditingThreshold) {
            ThresholdSheet(
                currentMinor: model.state.thresholdMinor,
                onSave: { minor in
                    model.setThreshold(minor: minor)
                    isEditingThreshold = false
                },
                onReset: {
                    model.clearThreshold()
                    isEditingThreshold = false
                }
            )
        }
    }

    // MARK: - The balance

    /// The wireframe's centred *"Balance (read-only)"* and its display figure.
    ///
    /// `balance`, not `available`: US-9.7 calls this the balance and D2' marks it read-only. The
    /// spendable figure is what every *decision* in this group is checked against — see
    /// ``WalletState/isBelowDayFee`` and ``CreditTransferModel/rejectionForSend()`` — and it is shown as
    /// its own line only when the two differ, because a driver with no accrued debt should not be
    /// reading two numbers for one wallet.
    private var balance: some View {
        VStack(spacing: 1) {
            Text(key: "wallet_balance_label")
                .mageFont(.caption)
                .foregroundStyle(MageRideColor.onSurfaceVariant)
            Text(model.state.balanceMinor.map { MoneyFormat.rupees($0) } ?? MoneyFormat.empty)
                .mageFont(.display)
                .foregroundStyle(MageRideColor.onSurface)

            if let debt = model.state.outstandingDebtMinor {
                Text("wallet_available_after_debt".localisedFormat(MoneyFormat.rupees(debt)))
                    .mageFont(.caption)
                    .foregroundStyle(MageRideColor.warning)
                    .multilineTextAlignment(.center)
            }
        }
        .frame(maxWidth: .infinity)
        .accessibilityElement(children: .combine)
    }

    // MARK: - The daily fee

    /// The wireframe's `Daily fee` card — vehicle, rate, and today's status (US-9.1/9.7).
    private var dailyFeeCard: some View {
        VStack(alignment: .leading, spacing: MageRideSpacing.xxs) {
            SectionLabel(key: "wallet_daily_fee")

            if let fee = model.state.standing.dailyFee {
                FeeRow(labelKey: "wallet_fee_vehicle", value: fee.vehicleType.labelKey.localised)
                FeeRow(
                    labelKey: "wallet_fee_rate",
                    value: "wallet_fee_rate_value".localisedFormat(
                        model.state.standing.dailyRateMinor.map { MoneyFormat.rupees($0) } ?? MoneyFormat.empty
                    )
                )
                FeeRow(labelKey: "wallet_fee_today", value: feeStatus, accent: feeAccent)
            } else {
                Text(key: "wallet_fee_unavailable")
                    .mageFont(.bodySmall)
                    .foregroundStyle(MageRideColor.onSurfaceVariant)
            }
        }
        .padding(MageRideSpacing.sm)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(
            MageRideColor.surfaceVariant,
            in: RoundedRectangle(cornerRadius: MageRideRadius.md, style: .continuous)
        )
    }

    /// *"PAID ✓ (first trip free)"* is two facts in one line, and the qualifier only belongs on a day
    /// the free trip has not been spent — see ``WalletFeeStanding/isFirstTripStillFree``.
    private var feeStatus: String {
        let status = (model.state.standing.isFeePaid ? "wallet_fee_paid" : "wallet_fee_unpaid").localised
        guard model.state.standing.isFirstTripStillFree else { return status }
        return "wallet_fee_status_first_free".localisedFormat(status)
    }

    private var feeAccent: Color {
        model.state.standing.isFeePaid ? MageRideColor.success : MageRideColor.warning
    }

    // MARK: - The banners

    /// The three banner states, ranked hardest first.
    ///
    /// Overdrawn is D5' §9.4's *"Top Up Required"*, *"below one day's fee"* is D2' §SCR-DI-021's, and
    /// the low-balance nudge is the softest of the three — see ``WalletState/isBelowDayFee`` on why the
    /// first two are different questions rather than one restated.
    @ViewBuilder
    private var notice: some View {
        if let owed = model.state.overdrawnByMinor {
            DashboardBanner(
                text: "wallet_top_up_required".localisedFormat(MoneyFormat.rupees(owed)),
                accent: MageRideColor.error,
                symbolName: "exclamationmark.triangle.fill"
            )
        } else if model.state.isBelowDayFee {
            DashboardBanner(
                text: "wallet_below_day_fee".localisedFormat(
                    model.state.standing.dailyRateMinor.map { MoneyFormat.rupees($0) } ?? MoneyFormat.empty
                ),
                accent: MageRideColor.error,
                symbolName: "exclamationmark.triangle.fill"
            )
        } else if model.state.isLowBalance {
            DashboardBanner(
                text: "wallet_low_balance".localisedFormat(MoneyFormat.rupees(model.state.thresholdMinor)),
                accent: MageRideColor.warning,
                symbolName: "exclamationmark.circle"
            )
        }
    }

    // MARK: - The actions

    /// The wireframe's two `btn-row`s — one CTA and three outlined actions.
    private var actions: some View {
        VStack(spacing: MageRideSpacing.xs) {
            Button { onOpen(.walletTopUp) } label: {
                Text(key: "wallet_action_top_up")
            }
            .buttonStyle(.mageCta)

            HStack(spacing: MageRideSpacing.xs) {
                OutlinedAction(labelKey: "wallet_action_request") { onOpen(.walletRequestCredit) }
                OutlinedAction(labelKey: "wallet_action_transfer") { onOpen(.walletTransfer) }
            }
            OutlinedAction(labelKey: "wallet_action_history") { onOpen(.walletHistory) }
        }
    }

    /// *"Warn me below Rs 200"* — the driver's own low-balance line (US-9.9).
    private var thresholdRow: some View {
        Button { isEditingThreshold = true } label: {
            HStack(spacing: MageRideSpacing.xxs) {
                Image(systemName: "bell.badge")
                    .font(.caption)
                Text("wallet_threshold_row".localisedFormat(MoneyFormat.rupees(model.state.thresholdMinor)))
                    .mageFont(.label)
                Spacer(minLength: 0)
            }
            .foregroundStyle(MageRideColor.primary)
            .frame(minHeight: MageRideControl.minimumTapTarget)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityElement(children: .combine)
        .accessibilityAddTraits(.isButton)
    }
}

/// The wireframe's `kv` — a muted label on the left, the value on the right.
private struct FeeRow: View {

    let labelKey: String
    let value: String
    var accent: Color?

    var body: some View {
        HStack(alignment: .firstTextBaseline, spacing: MageRideSpacing.xs) {
            Text(key: labelKey)
                .mageFont(.bodySmall)
                .foregroundStyle(MageRideColor.onSurfaceVariant)
                .frame(maxWidth: .infinity, alignment: .leading)
            Text(value)
                .mageFont(.bodySmall)
                .foregroundStyle(accent ?? MageRideColor.onSurface)
        }
        .accessibilityElement(children: .combine)
    }
}

/// The wireframe's `.btn-out` — an outlined action beside or under the CTA.
///
/// Not `.bordered`: §0.2 fixes the CTA's height and radius and this is the same bar without the fill,
/// which is what keeps the four buttons on this screen one control at four widths.
struct OutlinedAction: View {

    let labelKey: String
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Text(key: labelKey)
                .mageFont(.subtitle)
                .foregroundStyle(MageRideColor.primary)
                .frame(maxWidth: .infinity, minHeight: MageRideControl.ctaHeight)
                .overlay {
                    RoundedRectangle(cornerRadius: MageRideControl.ctaRadius, style: .continuous)
                        .strokeBorder(MageRideColor.outline, lineWidth: 1)
                }
                .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
    }
}

/// The threshold sheet.
///
/// The body says the setting lives on this handset, because it does and because a driver who believed
/// they had changed when MageRide texts them would be misled — the `LOW_BALANCE` push runs on the
/// admin's figure and there is no route that would let this one reach it.
///
/// A `.sheet` with `.presentationDetents([.medium])` rather than an `.alert`: an `.alert` on this
/// platform hosts at most one text field and gives it neither a label nor a `Rs` prefix, and this one
/// needs the explanation above it to be readable.
private struct ThresholdSheet: View {

    let currentMinor: Int64
    let onSave: (Int64) -> Void
    let onReset: () -> Void

    @Environment(\.dismiss) private var dismiss
    @State private var typed: String

    init(currentMinor: Int64, onSave: @escaping (Int64) -> Void, onReset: @escaping () -> Void) {
        self.currentMinor = currentMinor
        self.onSave = onSave
        self.onReset = onReset
        _typed = State(initialValue: WalletInput.rupeesOf(currentMinor))
    }

    var body: some View {
        NavigationStack {
            VStack(alignment: .leading, spacing: MageRideSpacing.sm) {
                Text(key: "wallet_threshold_body")
                    .mageFont(.bodySmall)
                    .foregroundStyle(MageRideColor.onSurfaceVariant)

                RupeeField(labelKey: "wallet_amount_label", value: $typed)

                Button(action: onReset) {
                    Text(key: "wallet_threshold_reset")
                        .mageFont(.label)
                        .foregroundStyle(MageRideColor.primary)
                }
                .frame(minHeight: MageRideControl.minimumTapTarget)

                Spacer(minLength: 0)

                Button { WalletInput.amountMinor(typed).map(onSave) } label: {
                    Text(key: "action_save")
                }
                .buttonStyle(.mageCta)
                .disabled(WalletInput.amountMinor(typed) == nil)
            }
            .padding(MageRideSpacing.md)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(MageRideColor.background)
            .navigationTitle(Text(key: "wallet_threshold_title"))
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button { dismiss() } label: { Text(key: "action_cancel") }
                }
            }
        }
        .presentationDetents([.medium])
    }
}
